### PR TITLE
Parse NSError to check if we should return NetworkError or not HTTPClientError

### LINKED ISSUE #26: BothamAPIClient should parse connection errors.
The current implementation is not parsing network errors coming from the ``HTTPClient`` implementation. We should detect the error code inside the NSError returned by the ``HTTPClient``implementation and map it to a ``BothamAPIClientError``.

### PR DESCRIPTION
**Pull request by pedrovgs**
Fix #26.

### CODE REVIEW

File: BothamNetworking/BothamAPIClient.swift [hunk 1]
```diff
@@ -115,6 +115,15 @@ public class BothamAPIClient {
         return interceptedResponse
     }
 
+    private func mapNSErrorToBothamAPIClientError(error: NSError) -> BothamAPIClientError {
```
Comment (davideme) [inline BothamNetworking/BothamAPIClient.swift]: We should to move this inside the implementation of NSHTTPClient, here we are exposing the implementation of NSHTTPClient and the error code that you should send in order to notify this kind of error.

File: BothamNetworking/BothamAPIClient.swift [hunk 2]
```diff
@@ -134,4 +143,8 @@ public class BothamAPIClient {
         return httpClient.hasValidScheme(request)
 
     }
+}
+
+private struct BothamAPIClientErrorCode {
+    private static let networkError = -1009
```
Comment (davideme) [inline BothamNetworking/BothamAPIClient.swift]: Take a look at the following page, there is constant for theses values, that have more explicit names.

http://nshipster.com/nserror/
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Miscellaneous/Foundation_Constants/index.html#//apple_ref/doc/constant_group/URL_Loading_System_Error_Codes

File: BothamNetworkingTests/ConnectionErrorHTTPClient.swift [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+//  Copyright © 2016 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import Result
+@testable import BothamNetworking
+
+class ConnectionErrorHTTPClient: HTTPClient {
+    func send(httpRequest: HTTPRequest, completion: (Result<HTTPResponse, NSError>) -> ()) {
+        let connectionError = NSError(domain: "NSURLErrorDomain", code: -1009, userInfo: nil)
```
Comment (davideme) [inline BothamNetworkingTests/ConnectionErrorHTTPClient.swift]: There is a constant for NSURLErrorDomain string.
https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Miscellaneous/Foundation_Constants/index.html#//apple_ref/doc/constant_group/NSURL_Domain

File: BothamNetworking/BothamAPIClient.swift (hunk 1 above)
Comment (pedrovgs) [inline BothamNetworking/BothamAPIClient.swift]: To do this, the best approach is to change HTTPClient interface to return ``Result<HTTPResponse, BothamAPIClientError>``. Is this your suggestion?

File: BothamNetworkingTests/ConnectionErrorHTTPClient.swift (hunk 3 above)
Comment (pedrovgs) [inline BothamNetworkingTests/ConnectionErrorHTTPClient.swift]: Fixed

File: BothamNetworking/BothamAPIClient.swift (hunk 2 above)
Comment (pedrovgs) [inline BothamNetworking/BothamAPIClient.swift]: Fixed

Comment (davideme) [inline BothamNetworking/BothamAPIClient.swift]: It look there is a bunch of others error code.
from -998 to -1021
Maybe we should we go the other way wrap error if it not returing a valid http status code.

Comment (pedrovgs) [inline BothamNetworking/BothamAPIClient.swift]: I don't have a status code there :( The only value I have to check is the error code. I'm going to review the errors and update the PR.

File: BothamNetworking/BothamAPIClient.swift (hunk 1 above)
Comment (davideme) [inline BothamNetworking/BothamAPIClient.swift]: We are already requesting to mapping HTTPRequest and HTTPResponse why not the Error.

Comment (pedrovgs) [inline BothamNetworking/BothamAPIClient.swift]: Done!

File: BothamNetworking/BothamAPIClient.swift (hunk 2 above)
Comment (pedrovgs) [inline BothamNetworking/BothamAPIClient.swift]: Done!

**Comment by pedrovgs** [comment]
@davideme I've reviewed all your comments and applied your suggestions. The major change is related to HTTPClient protocol which now returns ``BothamAPIClientError`` instead of NSError. Thanks for your feedback!

**Comment by davideme** [comment]
:+1:

**Pull request merged**